File: LeetCode/Difficulty/Medium/js/117.populating-next-right-pointers-in-each-node-II.js
/**
 * // Definition for a Node.
 * function Node(val, left, right, next) {
 *    this.val = val === undefined ? null : val;
 *    this.left = left === undefined ? null : left;
 *    this.right = right === undefined ? null : right;
 *    this.next = next === undefined ? null : next;
 * };
 */

/**
 * @param {Node} root
 * @return {Node}
 */
var connect = function (root) {
    const stack = runBFS(root);
    console.log(stack);

    let prev = null;
    while (stack.length > 0) {
        const top = stack.pop();
        if (prev && (top.level === prev.level)) {
            top.node.next = prev;
            prev = top;
        } else {
            prev = top;
        }
    }

    return root;
};

function runBFS(root) {
    const queue = []
        , stack = [];
    queue.push({
        node: root,
        level: 0
    });
    while (queue.length > 0) {
        const topNode = queue.shift();
        if (topNode.node && topNode.node.left)
            queue.push({
                node: topNode.node.left,
                level: topNode.level + 1
            });
        if (topNode.node && topNode.node.right)
            queue.push({
                node: topNode.node.right,
                level: topNode.level + 1
            });
        stack.push(topNode);
    }

    return stack;
}

File: LeetCode/Difficulty/Medium/cpp/117.populating-next-right-pointers-in-each-node-ii.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
    
    Node* getLeftMostChild(Node* node) {
        while (node) {
            if (node->left) return node->left;
            if (node->right) return node->right;
            node = node->next;
        }
        return NULL;
    }
    
public:
    Node* connect(Node* root) {
        for (Node* curr = root; curr; curr = getLeftMostChild(curr)) {
            for (Node* temp = curr; temp; temp = temp->next) {
                if (temp->left)
                    temp->left->next = temp->right ? temp->right : getLeftMostChild(temp->next);
                if (temp->right)
                    temp->right->next = getLeftMostChild(temp->next);
            }
        }
        return root;
    }
};